dds/DCPS/DataWriterImpl_T.h

```diff
@@ -341,82 +341,40 @@ DataWriterImpl_T
-     * encoding modes. We do this by picking the first supported representation
-     * that is supported by the mode, which is sorta what the XTypes spec says
-     * to do, except it only assumes one mode. One mode (but not both) may be
-     * left set to KIND_UNKNOWN, disabling it and preventing it from matching
-     * with readers just available through that mode.
-     */
-    Encoding::Kind encap_kind = Encoding::KIND_UNKNOWN;
-    Encoding::Kind nonencap_kind = Encoding::KIND_UNKNOWN;
-    DDS::DataRepresentationIdSeq repIds =
-      get_effective_data_rep_qos(qos_.representation.value);
-    for (CORBA::ULong i = 0; i < repIds.length(); ++i) {
-      const DDS::DataRepresentationId_t repr = repIds[i];
-      const Encoding::Kind kind =
-        repr_ext_to_encoding_kind(repr, MarshalTraitsType::extensibility());
-      switch (kind) {
-      case Encoding::KIND_CDR_UNALIGNED:
-        if (nonencap_kind == Encoding::KIND_UNKNOWN) {
-          nonencap_kind = kind;
+    DDS::DataRepresentationIdSeq repIds = get_effective_data_rep_qos(qos_.representation.value);
+    bool success = false;
+    Encoding::Kind encoding_kind;
+    if (cdr_encapsulation()) {
+      for (CORBA::ULong i = 0; i < repIds.length(); ++i) {
+        // TODO(sonndinh): Do we need to return failure when encountering an invalid representation
+        if (repr_to_encoding_kind(repIds[i], encoding_kind) &&
+            (Encoding::KIND_XCDR1 == encoding_kind ||
+             Encoding::KIND_XCDR2 == encoding_kind)) {
+          encoding_mode_ = EncodingMode(encoding_kind, swap_bytes());
+          success = true;
+          break;
         }
-        break;
-
-      case Encoding::KIND_CDR_PLAIN:
-      case Encoding::KIND_CDR_PARAMLIST:
-      case Encoding::KIND_XCDR2_PLAIN:
-      case Encoding::KIND_XCDR2_PARAMLIST:
-      case Encoding::KIND_XCDR2_DELIMITED:
-        if (encap_kind == Encoding::KIND_UNKNOWN) {
-          encap_kind = kind;
-        }
-        break;
-
-      case Encoding::KIND_XML:
-        // Not Supported, Ignore
-        break;
-
-      case Encoding::KIND_UNKNOWN:
-        if (::OpenDDS::DCPS::DCPS_debug_level) {
-          ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: ")
-            ACE_TEXT("%CDataWriterImpl::setup_serialization: ")
-            ACE_TEXT("Encountered unknown DataRepresentationId_t value: %d\n"),
-            TraitsType::type_name(), repr));
-        }
-        return DDS::RETCODE_ERROR;
       }
-      if (encap_kind != Encoding::KIND_UNKNOWN &&
-          nonencap_kind != Encoding::KIND_UNKNOWN) {
-        break;
+    } else { // Non-encapsulation
+      for (CORBA::ULong i = 0; i < repIds.length(); ++i) {
+        if (repr_to_encoding_kind(repIds[i], encoding_kind) &&
+            Encoding::KIND_UNALIGNED_CDR == encoding_kind) {
+          encoding_mode_ = EncodingMode(encoding_kind, swap_bytes());
+          success = true;
+          break;
+        }
```

Comment:
So the problem with this (and the code from before) is tied to the "How does Data Representation QoS and Annotations work with Unaligned CDR?" ticket, which is deciding on how much to involved unaligned CDR in the XTypes system. Right now the default representation QoS is just XCDR, which is what the spec says it should be, but that doesn't match how OpenDDS actually works. This means this code breaks all the tests that require unaligned CDR as none of them specify they need it. This is the reason the code from before was wrapped in a `#if 0`, so the tests wouldn't break when merged with #1688.

For now just change it to always be `Encoding::KIND_UNALIGNED_CDR` if there's no encapsulation.

Reply:
It is using unaligned CDR without iterating the list of data representations now.